### ISSUE TITLE
Improve progress bar design

### ISSUE THREAD

**Issue by DragaDoncila** [napari/napari#3002]
## 🚀 Feature
Following on from #2656 I wanted to document the currently outstanding design improvements suggested by @jni, @tlambert03, @goanpeca and me. I don't know if all of these suggestions *should* be implemented or whether they're the best options, so I think it would be good to get a design audit from @liaprins-czi before we go much further.

- [ ] **Fix Vertical Jumping**: as progress bars are added and closed, the displayed progress bars jump around to fill the available space. It would be nice to limit this and particularly for nested progress bars, reserve the rows of the child bars so that we don't get the jumping around.
- [ ] **Fix Horizontal Alignment**: progress bars contract when the ETA is added, and don't align depending on the size of the ETA label. The progress bar and ETA label should get fixed space to ensure they are aligned.
- [ ] **Improve Scrolling Behaviour**: the size of one scroll step is currently roughly equal to the height of a progress bar so there is a "strobing" effect when scrolling. We should consider ways to avoid this.
- [ ] **Remove Text Label**: Having the word `activity` next to the expand/contract arrow is not the prettiest. We could consider having a greyed out version of the in-progress indicator as a clickable button for expanding and contracting, or perhaps some other more indicative icon. The concerns here are mainly around discoverability so it would be good to get @liaprins-czi's ideas on this
- [ ] **Auto-Close Dock**: Consider automatically closing the activity dock when all the active progress bars have finished. Depending on the user workflow this could be pretty annoying as a new progress bar might start up pretty quickly after the last ones finished. Again, something @liaprins-czi should weigh in on imo.
- [ ] **Move Progress Indicator to Own Thread**: operations which block the main thread stop the progress indicator from moving smoothly and can be jarring. Having the indicator in its own thread could make for a much nicer experience.

I think I've covered all the current suggestions for improvement but of course any new suggestions would be welcome to make the progress bars as useful, discoverable and pretty as possible.

**Comment by jni** [comment]
> Depending on the user workflow this could be pretty annoying as a new progress bar might start up pretty quickly after the last ones finished

Yeah I thought about workflows finishing and immediately starting up again, and I agree it's an issue. Two options there would be:
- instruct plugin developers/users to always use nested progress when it makes sense
- have a timeout on the closing, so that the activity dock auto closes only after x seconds of inactivity.

**Comment by liaprins-czi** [comment]
Thanks for documenting all this and looping me in @DragaDoncila ! I have read through this issue and #2656 and all the options / issues documented here, very thorough! To do designs, I will need to know a bit more about context in terms of when progress bars appear and what they are communicating, context about any additional information shown, etc. Probably would be best to hae a call with @DragaDoncila (and anyone else involved / interested) to review and get up to speed. @sofroniewn I am curious how / if this work (for design improvements) fits into the "tablestakes" work, and what priority it is?

**Comment by goanpeca** [comment]
Thanks for the comments @DragaDoncila 

>Move Progress Indicator to Own Thread: operations which block the main thread stop the progress indicator from moving smoothly and can be jarring. Having the indicator in its own thread could make for a much nicer experience.

All GUI in Qt operations need to live on the main thread so I do not understand what the comment means exactly.

**Comment by goanpeca** [comment]
>have a timeout on the closing, so that the activity dock auto closes only after x seconds of inactivity.

I like this idea @jni  :)

**Comment by tlambert03** [comment]
> All GUI in Qt operations need to live on the main thread so I do not understand what the comment means exactly.

yeah... by it's very nature, the progress bar is a main-thread object relaying the status of (hopefully) some "other-thread" process

**Comment by sofroniewn** [comment]
> I am curious how / if this work (for design improvements) fits into the "tablestakes" work, and what priority it is?

This is part of "tablestakes", and I'd put this at pretty high priority @liaprins-czi.

I'm adding a screenshot here of latest

<img width="1289" alt="Screen Shot 2021-07-08 at 11 15 58 PM" src="https://user-images.githubusercontent.com/6531703/125128164-7ebc4400-e0b2-11eb-99da-7c877a6c1c85.png">

I wonder in general if our activity dialog/ error dialog should also be fixed to the bottom right of the viewer and not the bottom right of the canvas? Maybe unclear what to do if console is open etc. but I feel it looks out of place where it is now that i have a dock widget on the right.

> Fix Vertical Jumping

Agreed, I'd also like the vertical height to be smaller here and then grow only as needed.

> Remove Text Label: Having the word activity next to the expand/contract arrow is not the prettiest

Is the same dialog that contains out error messages? If so then only the arrow might make sense.

It might also be worth revisiting designs for #2290 by @Carreau at a similar time @liaprins-czi, which could contain some concepts we want to reuse

**Comment by DragaDoncila** [comment]
> Probably would be best to hae a call with @DragaDoncila (and anyone else involved / interested) to review and get up to speed. 

Completely agree @liaprins-czi , let's set something up for next week.

**Comment by liaprins-czi** [comment]
While I start working on the designs for this issue, I have made a document to keep track of design considerations:
https://docs.google.com/document/d/1uIf6OAPuaDOd0guVUCc2pTgRC6iV_u9H7q1bTmqLKhI/edit?usp=sharing

I have highlighted some of the considerations that are more like questions I have at this point, that the design explorations and decisions will be dependent upon. If anyone has answers, it'd be great to have your comments in teh doc (@DragaDoncila maybe?)! Thanks!

**Comment by liaprins-czi** [comment]
Here's the Figma where I've started explorations + grouped otpions for feedback by the progress-bar-related issues they attempt to solve:
https://www.figma.com/file/nl19TPV1yOuoCBVfgLGpAS/Progress-bars?node-id=36%3A1580

**Comment by isabela-pf** [comment]
Hi there everyone! I presented a selection of @liaprins-czi’s mockups at the community call on October 7 and we discussed preferences about a few aspects of these designs so we have a clear path for the first iteration or implementation. I’m going to share the exported version of the comparisons here and summarize the comments made.

# Questions of placement

## Status bar or side panel
Do progress bars belong in pop-ups tied to the status bar, or in the left side panel? Mock up shown:

<img width="2780" alt="" src="https://user-images.githubusercontent.com/50221806/136492148-6d35b1da-f621-4369-8066-0798048853d9.png">

**There was a preference for the pop-up** over the canvas approach, though it was noted people would like a smaller default size than was shown in the above mock up. We will have to test how this will work with other napari notifications since they will now be taking the same place on the screen; the basic goal should be to make sure they simply stack on top of one another.

As a future note, this may be worth revisiting if/when napari holds notification history. Over the canvas feels like a much more transient place than the sidebar.

### Fixed or resized pop-ups
If progress bars do belong in pop-ups, is the pop-up sized fixed or can it be controlled (resized) by the user? Mock up shown:

<img width="2623" alt=" " src="https://user-images.githubusercontent.com/50221806/136492167-00d6ed96-cdd7-4363-92c1-2a3ef3fc30e5.png">

There was more variety in this discussion, but in the end there seemed to be **more interest in having a fixed size**. A few people mentioned that because there is a collapsed state for progress bars, it seems less critical to give users the ability to resize the progress bar area. Having a small fixed size that works even for the minimum window size of napari will be important here.

## Persistent or transient progress bar UI
Do progress bars UI elements only appear when there are tasks running, or are they always in the interface? Mock up shown:

<img width="900" alt="" src="https://user-images.githubusercontent.com/50221806/136492203-ed860a8d-bf3b-4230-a2ca-23a838cc6eb4.png">

**Transient UI—meaning the icon for progress bars will not appear with no tasks running—was the choice here.** 

Specifically, this works because there is nothing else in the lower right of the status bar at the moment, so it is not interfering with other UI when it appears and disappears. If there are other UI elements in this area in the future, we should consider having this be a persistent button disabled when no tasks are running.

# Questions of representation

## Collapsed task UI
Are tasks better represented as dots or a single spinner? Mock up shown:

<img width="900" alt="" src="https://user-images.githubusercontent.com/50221806/136492232-493d27fe-7d78-456b-ae97-aa21576f346a.png">

Both approaches had distinct pros and cons. The dots seem like they provide the most useful information in a glance, but will not be as helpful for a large number of tasks. The spinner does not give specific information as to the number of tasks, but it will be easy to tell when something is running versus it being stuck. There was also interest in seeing a combined approach with the dots each serving as their own small progress bar.

**The final decision was to start with the dots** since they seem easier to implement. Next step goal is to have a circular progress bar that accurately reflects the task being run.

Side note: we will most likely need to be thinking about an approach for overflowing elements in the status bar. There are a few ideas for this I can post later on.

## Highest level tasks only or nested tasks
Do progress bars show only the highest level tasks, or do they show children of the highest-level tasks? Mock up shown:

<img width="1722" alt="s" src="https://user-images.githubusercontent.com/50221806/136492255-81384e60-3623-422f-afba-4988c90b7585.png">

**People agreed they should be able to show children of tasks in the progress bar. The segmented progress bar (like multiple progress bars in a row) was also the most popular representation for this.**

### If there are children progress bars, how many levels is helpful? 
**One or two levels of progress bars** would be helpful. There was also a request that the children be collapsible (what a silly statement out of context).


---

Stay tuned for an update on how all these choices will look together in a simple prototype.

As mentioned above, these are choices for the first attempt. Feedback is always welcome, whether that’s for the first round or something you’d like to see as the progress bars develop. Please let me know if you have any additional thoughts or think I’ve missed something from the meeting.

**Comment by sofroniewn** [comment]
Wow! I love these side by sides so much!! Having missed the meeting the other day they make it so easy for me to catch up and see what was compared and what was decided. All the choices sound great to me and the current plan seems like a great place to start. Nice work 🚀

**Comment by goanpeca** [comment]
These mockups look great! I kinda like the spinner approach, could we perhaps include a number within the spinner so that it give information on tasks being worked on and maybe the number of them @isabela-pf ?

**Comment by isabela-pf** [comment]
Good question @goanpeca! That would combine the best of both options well. @liaprins-czi is super thorough and made several mock ups trying out this approach (like this once below).

<img width="692" alt="Activity icon when widget open" src="https://user-images.githubusercontent.com/50221806/137192057-c5d57634-4384-4eff-97e0-6657f30b055d.png">

I chose not to include her work with those options when I shared at the community call because I think it won't hold up well to how many tasks a user could queue up at once. I made some examples to show what I mean. So it looks really good with single digit numbers as shown above, right?

It starts to get a little tight with double digits, but still works.
<img width="376" alt="Screen Shot 2021-10-13 at 11 32 26 AM" src="https://user-images.githubusercontent.com/50221806/137192658-f24a040d-399f-4fd9-aa1a-4f2d8ed1ac99.png">

But it can't fit in the circle anymore once we hit triple-digit numbers. And having tasks in the triple digits (or more) isn't impossible, so we can't ignore it. 
<img width="423" alt="Why Isabela didn't include the number and spinner combo" src="https://user-images.githubusercontent.com/50221806/137192979-4eced41b-08ff-42e7-a68b-ba6bd5a56014.png">

We'd have to agree on some kind of design for what to do when we have three or more digits to account for, and I don't think that's a problem we need to make more complicated for ourselves. We do still have to solve this for the aforementioned "dot approach" as well, but we have more freedom to use the space when we aren't tied to fitting in or around a single circle.

Hope that was helpful info! Let me know if you have any other questions or people think the combined approach is worth working out.

**Comment by goanpeca** [comment]
Very helpful! Just what I had in mind. I don't think three digits are that common 🤔 but maybe they do... maybe we have a `...` for those cases and on hover more info is displayed on tooltip 🤷🏼 

Awesome mocks :) Thanks for the clarification

**Comment by DragaDoncila** [comment]
I think the overall concern we had with the spinner is that it's indeterminate (just spinning to indicate activity rather than progress to completion). We could probably do something to roughly assess progress to completion, but it would be across all progress bars rather than per bar. The "per bar" information is what I think attracted us to the dots approach, in particular, to a dots approach where the circle fills in proportionally to the bar. Even without a per bar proportionate circle, we can signal more clearly when a new progress bar begins/ends with the dots approach.

**Comment by isabela-pf** [comment]
Based on last week's community feedback, here's [the promised (tiny) prototype](https://www.figma.com/proto/nl19TPV1yOuoCBVfgLGpAS/Progress-bars?page-id=10%3A2639&node-id=254%3A2004&viewport=288%2C48%2C0.06&scaling=scale-down&starting-point-node-id=254%3A2004). If you click the button to collapse the pop up progress bars on the first page, the rest will progress on its own.

This is the combined UI elements discussed above in one image (the first image of the prototype):
<img width="1277" alt="napari viewer with a pop up window holding four progress bars." src="https://user-images.githubusercontent.com/50221806/137437403-766a6ca3-0232-49a3-ad47-8c8b55dda9db.png">

This is what it looks like in the minimum window size napari has. It's what I based the fixed size window on the other mock ups as well.
<img width="614" alt="A small napari window with the progress bars pop up in it." src="https://user-images.githubusercontent.com/50221806/137437598-d4d69f77-83d8-4efd-ab39-79d4f2c2e836.png">

There are definitely more interactions to work through, but this should help give an idea as to how all the above decisions play together. As always, let me know what you think!

Some particular items I'm undecided on:
- the pop up size
- the number of dots (three in this case) shown before going to the overflow description
- the "task complete" green dot and checkmark

**Comment by liaprins-czi** [comment]
I love how this is shaping up! Great to be getting to consolidated feedback and decisions! 🎉

**Comment by DragaDoncila** [comment]
@isabela-pf just want to clarify as well that currently the dialog is pinned to the bottom-right corner of the canvas *not* the window, so when the console is open the activity dialog doesn't cover it. I don't know if we want to keep it that way - my intuition is yes

**Comment by isabela-pf** [comment]
@DragaDoncila yes! That was my mistake in mocking it up; thanks for catching that.

**Comment by isabela-pf** [comment]
For next steps, I was thinking of all that could go wrong (as I do) and trying to design for progress bars breaking. It seems like It'd be very possible to have a task fail to complete, and so we'd need to communicate that to users in some way.

I just started exploring ideas for this ([on figma](https://www.figma.com/file/nl19TPV1yOuoCBVfgLGpAS/Progress-bars?node-id=265%3A2283)), but immediately realized that this may not work with the transient nature of tasks/progress bars as we are planning to have them now. Does it make sense to give feedback on any errors here, or is it all just in the logs?

**Comment by DragaDoncila** [comment]
> It seems like It'd be very possible to have a task fail to complete, and so we'd need to communicate that to users in some way.

Hmm yes... 😆 . So at the moment, if a progress bar fails, the exception is caught by our notification manager and the associated progress bar disappears - the same way as it does when it completes. 

> Does it make sense to give feedback on any errors here, or is it all just in the logs?

We don't yet have a concept of "logs" - notifications are popped up to the user, but if dismissed they aren't stored anywhere for reference. 
I think this goes back to the discussion about whether we want completed progress bars to stick around for any amount of time. Currently they just disappear as soon as they hit 100%. 
If they were to stick around until explicit dismissal OR for a set amount of time, then I think the fact that it errored should be displayed - but is that very useful if we can't tell them why/how it failed?
If we keep the current behavior of disappearing on completion, then I don't think we should do anything different when it errors.

@isabela-pf I don't know if this answers your question or not. Maybe this is something we should bring back to the community - "do we want progress bars to stick around for any amount of time after completion/error or do we want to maintain existing behavior"?

**Comment by isabela-pf** [comment]
@DragaDoncila thanks for following up on this and answering my question. This is a good point in favor of having the (more complicated) progress bar experience where they persist, though it definitely has more details to work out. I don't think this should be a blocker, though, since having any kind of feedback on progress seems more helpful than we currently have.

I can put this on my list of things to bring up in a community meeting. Thanks!

**Comment by isabela-pf** [comment]
Update on persistent progress bars from the community meeting:

@DragaDoncila and I brought up the pros and cons of having a progress bar experience that is persistent/shows completed jobs. Overall the feedback was to start without that feature (since it brings up many implementation and design questions) and reassess it in the future. This seems to be more of a wish list item than a need.
